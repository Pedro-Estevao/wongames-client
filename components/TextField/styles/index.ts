'use client';

import styled, { css } from 'styled-components';
import { TextFieldProps } from '..';

type IconPositionProps = Pick<TextFieldProps, 'iconPosition'>;

type ContainerProps = Pick<TextFieldProps, 'disabled'> & { error?: boolean };

const containerModifiers = {
	error: () => css`
		${InputWrapper} {
			border-color: #ff6347;
		}

		${Icon},
		${Label} {
			color: #ff6347;
		}
	`,
	disabled: () => css`
		${Label},
		${Input},
        ${Icon} {
			cursor: not-allowed;
			color: #8f8f8f;

			&::placeholder {
				color: currentColor;
			}
		}
	`
};

export const TextFieldContainer = styled.div<ContainerProps>`
	${({ disabled, error }) => css`
		${disabled && containerModifiers.disabled()}
		${error && containerModifiers.error()}
	`}
`;

export const InputWrapper = styled.div`
	${() => css`
		display: flex;
		background-color: #eaeaea;
		border-radius: 0.2rem;
		padding: 0 1.6rem;
		border: 0.2rem solid;
		border-color: #eaeaea;

		&:focus-within {
			box-shadow: 0 0 0.5rem #f231a5;
		}
	`}
`;

export const Input = styled.input<IconPositionProps>`
	${({ iconPosition }) => css`
		color: #030517;
		font-family: 'Poppins', sans-serif;
		font-size: 3.2rem;
		padding: 0.8rem 0;
        padding-${iconPosition}: 1.6rem;
		background: transparent;
		border: 0;
		outline: none;
		width: 100%;
	`}
`;

export const Label = styled.label`
	${() => css`
		font-size: 2.4rem;
		color: #030517;
		cursor: pointer;
	`}
`;

export const Icon = styled.div<IconPositionProps>`
	${({ iconPosition }) => css`
		display: flex;
		width: 2.2rem;
		color: #8f8f8f;
		order: ${iconPosition === 'right' ? 1 : 0};

		& > svg {
			width: 100%;
		}
	`}
`;

export const Error = styled.p`
	${() => css`
		color: #ff6347;
		font-size: 1.6rem;
	`}
`;
