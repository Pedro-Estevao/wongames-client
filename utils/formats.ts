export const formatPrice = (price: number | bigint) => {
	return new Intl.NumberFormat('en', {
		style: 'currency',
		currency: 'USD'
	}).format(price);
};

export const formatDate = (date: string) => {
	return new Intl.DateTimeFormat('en-US', {
		day: 'numeric',
		month: 'short',
		year: 'numeric'
	}).format(new Date(date));
};

export const formatImageUrl = (url: string | undefined) => {
	if (process.env.NEXT_PUBLIC_API_URL) {
		return `${process.env.NEXT_PUBLIC_API_URL}${url}`;
	}

	if (url) {
		return url;
	}

	return null;
};