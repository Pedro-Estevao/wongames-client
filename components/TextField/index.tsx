'use client';

import React, { forwardRef, useState } from 'react';
import * as S from './styles';

export type TextFieldProps = {
	onInputChange?: (value: string) => void;
	label?: string;
	initialValue?: string;
	icon?: React.ReactNode;
	iconPosition?: 'left' | 'right';
	disabled?: boolean;
	error?: string;
} & React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;

const TextField = forwardRef<HTMLInputElement, TextFieldProps>((
	{
		icon,
		iconPosition = 'left',
		label,
		name,
		initialValue = '',
		error,
		disabled = false,
		onInputChange,
		...props
	},
	ref
) => {
	const [value, setValue] = useState<string>(initialValue);

	const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		const newValue = e.currentTarget.value;
		setValue(newValue);

		if (onInputChange) {
			onInputChange(newValue);
		}
	};

	return (
		<S.TextFieldContainer $disabled={disabled} $error={!!error}>
			{!!label && <S.Label htmlFor={name}>{label}</S.Label>}
			<S.InputWrapper>
				{!!icon && <S.Icon $iconPosition={iconPosition}>{icon}</S.Icon>}
				<S.Input
					ref={ref}
					type="text"
					onChange={onChange}
					value={value}
					$iconPosition={iconPosition}
					disabled={disabled}
					name={name}
					{...(label ? { id: name } : {})}
					{...props}
				/>
			</S.InputWrapper>
			{!!error && <S.Error>{error}</S.Error>}
		</S.TextFieldContainer>
	);
});

TextField.displayName = 'TextField';

export default TextField;
