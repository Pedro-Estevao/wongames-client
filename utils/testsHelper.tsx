import { CartContext, CartContextData, CartContextDefaultValues } from "@/hooks/useCart";
import { WishlistContext, WishlistContextData, WishlistContextDefaultValues } from "@/hooks/useWishlist";
import { theme } from "@/public/styles/theme";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "styled-components";

type CustomRenderProps = {
    cartProviderProps?: CartContextData;
    wishlistProviderProps?: WishlistContextData;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
    ui: React.ReactElement,
    {
        cartProviderProps = CartContextDefaultValues,
        wishlistProviderProps = WishlistContextDefaultValues,
        ...renderOptions
    }: CustomRenderProps = {}
) => render(
    <ThemeProvider theme={theme}>
        <CartContext.Provider value={cartProviderProps}>
            <WishlistContext.Provider value={wishlistProviderProps}>
                {ui}
            </WishlistContext.Provider>
        </CartContext.Provider>
    </ThemeProvider>,
    renderOptions
);

export * from "@testing-library/react";
export { customRender as render };