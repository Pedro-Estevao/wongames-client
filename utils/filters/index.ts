import { ItemProps } from '@/components/ExploreSidebar';
import { FilterObject, FilterValue } from "@/components/Layouts/Games";
import { ParsedUrlQueryInput } from 'querystring';

type ParseArgs = {
	queryString: ParsedUrlQueryInput;
	filterItems: Pick<ItemProps, 'type' | 'name'>[];
};

export const parseQueryStringToFilter = ({
	queryString,
	filterItems
}: ParseArgs) => {
	const obj: Record<string, string | string[]> = {};

	Object.keys(queryString).forEach((key) => {
		const item = filterItems?.find((item) => item.name === key);
		const isCheckbox = item?.type === 'checkbox';
		const queryValue = queryString[key];

		if (Array.isArray(queryValue)) {
			// Remove duplicatas do array
			obj[key] = Array.from(new Set(queryValue));
		} else if (typeof queryValue === 'string') {
			if (isCheckbox) {
				// Se for um checkbox, converte a string em array e remove duplicatas
				const values = queryValue.split(',').filter(Boolean); // Remove strings vazias
				obj[key] = Array.from(new Set(values));
			} else {
				// Se não for checkbox, mantém a string como está
				obj[key] = queryValue;
			}
		}
	});

	return obj;
};

export const parseQueryStringToObjFilter = (
	queryString: Record<string, string | string[]>,
	filterItems: ItemProps[]
): FilterObject => {
	const filters: FilterObject = {};

	filterItems.filter((item) => item.name !== 'sort').forEach((item) => {
		if (queryString[item.name]) {
			const queryValue = queryString[item.name];
			let filterValue: FilterValue | string | string[] | number;

			if (Array.isArray(queryValue)) {
				filterValue = queryValue;
			} else {
				// Check if the value is a number (used for price)
				filterValue = isNaN(Number(queryValue)) ? queryValue : Number(queryValue);

				if (item.type === 'checkbox' && typeof filterValue === 'string') {
					filterValue = filterValue.split(',').map(str => str.trim());
				}
			}

			filters[item.name] = item.value
				? item.filter
					? { [item.value]: { [item.filter]: filterValue } }
					: { [item.value]: filterValue }
				: item.filter
					? { [item.filter]: filterValue }
					: filterValue;
		}
	});

	return filters;
};