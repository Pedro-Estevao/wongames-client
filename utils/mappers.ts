import { BannerProps } from '@/components/Banner';
import { GameCardProps } from '@/components/GameCard';
import { HighlightProps } from '@/components/Highlight';
import {
	BannerEntity,
	BannerEntityResponseCollection,
	BannerFragmentFragment,
	GameEntityResponseCollection,
	GameFragmentFragment,
	HighlightFragmentFragment
} from '@/graphql/generated/graphql';

export const bannerMapper = (banners: BannerEntityResponseCollection) => {
	return (
		(banners?.data?.map((banner: BannerEntity) => ({
			img: `${process.env.NEXT_PUBLIC_API_URL}${(
				banner.attributes as BannerFragmentFragment
			)?.image?.data?.attributes?.url}`,
			title: (banner.attributes as BannerFragmentFragment)?.title || '',
			subTitle: (banner.attributes as BannerFragmentFragment)?.subTitle || '',
			buttonLabel:
				(banner.attributes as BannerFragmentFragment)?.button?.label || '',
			buttonLink:
				(banner.attributes as BannerFragmentFragment)?.button?.link || '',
			...((banner.attributes as BannerFragmentFragment)?.ribbon && {
				ribbon:
					(banner.attributes as BannerFragmentFragment)?.ribbon?.text || null,
				ribbonColor:
					(banner.attributes as BannerFragmentFragment)?.ribbon?.color || null,
				ribbonSize:
					(banner.attributes as BannerFragmentFragment)?.ribbon?.size || null
			})
		})) as BannerProps[]) || []
	);
};

export const gamesMapper = (games: GameEntityResponseCollection) => {
	return (
		(games?.data.map((game) => ({
			title: (game.attributes as GameFragmentFragment).name,
			slug: (game.attributes as GameFragmentFragment).slug,
			developer: (game.attributes as GameFragmentFragment).developers?.data[0]
				.attributes?.name,
			img: `${process.env.NEXT_PUBLIC_API_URL}${(game.attributes as GameFragmentFragment)
				.cover?.data?.attributes?.url}`,
			price: (game.attributes as GameFragmentFragment).price
		})) as GameCardProps[]) || []
	);
};

export const staticGamesParams = (games: GameEntityResponseCollection) => {
	return (
		(games?.data.map((game) => ({
			title: (game.attributes as GameFragmentFragment).name,
			slug: (game.attributes as GameFragmentFragment).slug,
			developer: (game.attributes as GameFragmentFragment).developers?.data[0]
				.attributes?.name,
			img: `${process.env.NEXT_PUBLIC_API_URL}${(game.attributes as GameFragmentFragment)
				.cover?.data?.attributes?.url}`,
			price: (game.attributes as GameFragmentFragment).price
		})) as GameCardProps[]) || []
	);
};

export const highlightMapper = (highlight: HighlightFragmentFragment) => {
	return (
		highlight &&
		({
			title: highlight.title,
			subTitle: highlight.subTitle,
			backgroundImage: `${process.env.NEXT_PUBLIC_API_URL}${highlight.background?.data?.attributes?.url}`,
			floatImage: `${process.env.NEXT_PUBLIC_API_URL}${highlight.floatImage?.data?.attributes?.url}`,
			buttonLabel: highlight.buttonLabel,
			buttonLink: highlight.buttonLink,
			alignment: highlight.alignment
		} as HighlightProps)
	);
};
