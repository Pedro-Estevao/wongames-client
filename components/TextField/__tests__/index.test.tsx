import { render, screen, waitFor } from '@/utils/testsHelper';
import { expect } from '@jest/globals';
import TextField from "@/components/TextField";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faEnvelope } from '@fortawesome/free-solid-svg-icons';
import userEvent from "@testing-library/user-event";

describe('<TextField />', () => {
    it('Renders with Label', () => {
        render(<TextField label="Label" name="Label" />);

        expect(screen.getByLabelText('Label')).toBeInTheDocument();
    });

    it('Renders without Label', () => {
        render(<TextField />);

        expect(screen.queryByLabelText('Label')).not.toBeInTheDocument();
    });

    it('Renders with placeholder', () => {
        render(<TextField placeholder="hey you" />);

        expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument();
    });

    it('Renders with Icon', () => {
        render(<TextField icon={<FontAwesomeIcon icon={faEnvelope} data-testid="icon" />} />);

        expect(screen.getByTestId('icon')).toBeInTheDocument();
    });

    it('Renders with Icon on the right side', () => {
        render(
            <TextField icon={<FontAwesomeIcon icon={faEnvelope} data-testid="icon" />} iconPosition="right" />
        );

        // eslint-disable-next-line testing-library/no-node-access
        expect(screen.getByTestId('icon').parentElement).toHaveStyle({ order: 1 });
    });

    it('Changes its value when typing', async () => {
        const onInputChange = jest.fn();
        render(
            <TextField
                onInputChange={onInputChange}
                label="TextField"
                name="TextField"
            />
        );

        const input = screen.getByRole('textbox');
        const text = 'This is my new text';
        await userEvent.type(input, text);

        await waitFor(() => expect(input).toHaveValue(text));
        await waitFor(() => expect(onInputChange).toHaveBeenCalledTimes(text.length));
        expect(onInputChange).toHaveBeenCalledWith(text);
    });

    it('Does not changes its value when disabled', async () => {
        const onInputChange = jest.fn();
        render(
            <TextField
                onInputChange={onInputChange}
                label="TextField"
                name="TextField"
                disabled
            />
        );

        const input = screen.getByRole('textbox');
        expect(input).toBeDisabled();

        const text = 'This is my new text';
        await userEvent.type(input, text);

        await waitFor(() => {
            expect(input).not.toHaveValue(text)
        });
        expect(onInputChange).not.toHaveBeenCalled();
    });

    it('Renders with error', () => {
        const { container } = render(
            <TextField
                icon={<FontAwesomeIcon icon={faEnvelope} data-testid="icon" />}
                label="TextField"
                error="Error message"
            />
        );
        const { firstChild } = container;

        expect(screen.getByText('Error message')).toBeInTheDocument();

        expect(firstChild).toMatchSnapshot();
    });

    it('Is accessible by tab', async () => {
        render(<TextField label="TextField" name="TextField" />);

        const input = screen.getByLabelText('TextField');
        expect(document.body).toHaveFocus();

        await userEvent.tab();
        expect(input).toHaveFocus();
    });

    it('Is not accessible by tab when disabled', async () => {
        render(<TextField label="TextField" name="TextField" disabled />);

        const input = screen.getByLabelText('TextField');
        expect(document.body).toHaveFocus();

        await userEvent.tab();
        expect(input).not.toHaveFocus();
    });
});