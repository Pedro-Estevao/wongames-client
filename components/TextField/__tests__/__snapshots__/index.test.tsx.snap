// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Renders with error 1`] = `
.c0 .c3 {
  border-color: #FF6347;
}

.c0 .c5,
.c0 .c1 {
  color: #FF6347;
}

.c4 {
  display: flex;
  background-color: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c7 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c7:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #EAEAEA inset;
  filter: none;
}

.c7:-webkit-autofill::first-line {
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c6 {
  display: flex;
  align-items: center;
  width: 2.2rem;
  color: #8F8F8F;
  order: 0;
}

.c6>svg {
  width: 2.2rem;
  height: 100%;
}

.c8 {
  color: #FF6347;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
  >
    TextField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-envelope "
        data-icon="envelope"
        data-prefix="fas"
        data-testid="icon"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M48 64C21.5 64 0 85.5 0 112c0 15.1 7.1 29.3 19.2 38.4L236.8 313.6c11.4 8.5 27 8.5 38.4 0L492.8 150.4c12.1-9.1 19.2-23.3 19.2-38.4c0-26.5-21.5-48-48-48L48 64zM0 176L0 384c0 35.3 28.7 64 64 64l384 0c35.3 0 64-28.7 64-64l0-208L294.4 339.2c-22.8 17.1-54 17.1-76.8 0L0 176z"
          fill="currentColor"
        />
      </svg>
    </div>
    <input
      class="c7"
      type="text"
      value=""
    />
  </div>
  <p
    class="c8"
  >
    Error message
  </p>
</div>
`;
